synapse/handlers/ui_auth/checkers.py

```diff
@@ -237,11 +237,65 @@ async def check_auth(self, authdict: dict, clientip: str) -> Any:
         return await self._check_threepid("msisdn", authdict)
 
 
+class RegistrationTokenAuthChecker(UserInteractiveAuthChecker):
+    AUTH_TYPE = LoginType.REGISTRATION_TOKEN
+
+    def __init__(self, hs: "HomeServer"):
+        super().__init__(hs)
+        self._enabled = bool(hs.config.registrations_require_token)
+        self.store = hs.get_datastore()
+
+    def is_enabled(self) -> bool:
+        return self._enabled
+
+    async def check_auth(self, authdict: dict, clientip: str) -> Any:
+        if "token" not in authdict:
+            raise LoginError(400, "Missing token", Codes.MISSING_PARAM)
+        if not isinstance(authdict["token"], str):
+            raise LoginError(400, "Token must be a string", Codes.INVALID_PARAM)
+        if "session" not in authdict:
+            raise SynapseError(400, "Missing UIA session", Codes.MISSING_PARAM)
+        session = authdict["session"]
+
+        # If the LoginType.REGISTRATION_TOKEN stage has already been completed,
+        # return early to avoid incrementing `pending` again.
+        # TODO: check stored token matches provided token?
+        try:
+            stored_token = await self.store.get_ui_auth_session_data(
+                session,
+                "registration_token",
+                None,
+            )
+        except StoreError:
+            raise SynapseError(400, "Unknown session ID: {}".format(session))
+        if stored_token:
+            return True
+
+        token = authdict["token"]
+        if await self.store.registration_token_is_valid(token):
+            # Increment pending counter, so that if token has limited uses it
+            # can't be used up by someone else in the meantime.
+            await self.store.set_registration_token_pending(token)
```

Comment:
We'll need to do something about expiring UIAA sessions at some point, though I don't think there's currently a method of doing so in Synapse...

Reply:
I completely forgot about that...
I think UIA sessions expire after 2 days, but that might be a bit long.

https://github.com/matrix-org/synapse/blob/52c60bd0a96cd61583209a9ef6d8270425e8a902/synapse/handlers/auth.py#L187

Reply:
I wonder how problematic it would be to expire a UIAA stage before expiring the whole session.

We check for expired sessions every 5 minutes here:

https://github.com/matrix-org/synapse/blob/bd4919fb72b2a75f1c0a7f0c78bd619fd2ae30e8/synapse/handlers/auth.py#L264-L271

The registration flow on clients may become a bit confused if they note a completed token stage, only for it to be no longer completed later on...

Definitely not a problem we need to first in the first pass anyhow.

Reply:
Yeah, it looks a bit tricky...
I guess first thing is to make sure pending is actually decreased when the whole session expires at least

Reply:
Pending is decreased when uncompleted sessions expire in e7495e61a9a50ee49ac9e8259094444b1bf05b55, but that does take 2 days.

Reply:
Ok, I'll leave it like it is then and if it turns out to be a problem I can try figure something else out.